### PR TITLE
bmp280: SPI support

### PR DESCRIPTION
**Pull request by mlyle**
Also factor some variables more tightly.

### CODE REVIEW

**Comment by mlyle** [comment]
spurious

jenkins, test this please

File: flight/PiOS/Common/pios_bmp280.c [hunk 1]
```diff
@@ -200,6 +195,36 @@ int32_t PIOS_BMP280_Init(const struct pios_bmp280_cfg *cfg, int32_t i2c_device)
...
+		return -1;
+
+	dev->i2c_id = i2c_device;
+
+	return PIOS_BMP280_Init_Common(cfg);
+}
+#endif /* PIOS_EXCLUDE_BMP280_I2C */
+
+#ifdef PIOS_INCLUDE_BMP280_SPI
+int32_t PIOS_BMP280_Init_SPI(const struct pios_bmp280_cfg *cfg, int32_t spi_device,
```
Comment (tracernz) [inline flight/PiOS/Common/pios_bmp280.c]: I did `PIOS_*_SPI_Init` in pios_mpu. Seems better to me to keep the least general part of the symbol at the end. WDYT? Also applies to the other I2C/SPI specific functions.

**Comment by mlyle** [comment]
Missing some commit; deferring slightly

**Comment by mluessi** [comment]
Which target uses this?

**Comment by mlyle** [comment]
@mluessi-- messaged you on IRC about that.

Updated per review feedback and to fix build of it.  Ready for review.

File: flight/PiOS/Common/pios_bmp280.c [hunk 2]
```diff
@@ -115,11 +119,13 @@ static struct bmp280_dev *dev;
  */
 static struct bmp280_dev *PIOS_BMP280_alloc(void)
 {
-   struct bmp280_dev *bmp280_dev;
+	struct bmp280_dev *bmp280_dev;
 
 	bmp280_dev = (struct bmp280_dev *)PIOS_malloc(sizeof(*bmp280_dev));
 	if (!bmp280_dev) return (NULL);
```
Comment (tracernz) [inline flight/PiOS/Common/pios_bmp280.c]: Newline.

File: flight/PiOS/Common/pios_bmp280.c [hunk 3]
```diff
@@ -115,11 +119,13 @@ static struct bmp280_dev *dev;
...
 
+	bzero(bmp280_dev, sizeof(*bmp280_dev));
```
Comment (tracernz) [inline flight/PiOS/Common/pios_bmp280.c]: Tsk tsk. 😁

File: flight/PiOS/Common/pios_bmp280.c [hunk 4]
```diff
@@ -143,22 +149,11 @@ static int32_t PIOS_BMP280_Validate(struct bmp280_dev *dev)
...
-	if (dev->i2c_id == 0)
-		return -3;
 	return 0;
 }
 
-/**
- * Initialise the BMP280 sensor
- */
-int32_t PIOS_BMP280_Init(const struct pios_bmp280_cfg *cfg, int32_t i2c_device)
+static int32_t PIOS_BMP280_Common_Init(const struct pios_bmp280_cfg *cfg)
```
Comment (tracernz) [inline flight/PiOS/Common/pios_bmp280.c:160]: 👍

File: flight/PiOS/Common/pios_bmp280.c [hunk 5]
```diff
@@ -200,6 +195,36 @@ int32_t PIOS_BMP280_Init(const struct pios_bmp280_cfg *cfg, int32_t i2c_device)
 	return 0;
 }
 
+/**
+ * Initialise the BMP280 sensor
+ */
+#ifndef PIOS_EXCLUDE_BMP280_I2C
```
Comment (tracernz) [inline flight/PiOS/Common/pios_bmp280.c:206]: Seems weird to use EXCLUDE for one and INCLUDE for the other. :/ I know it saves you diff but... ;)

File: flight/PiOS/Common/pios_bmp280.c [hunk 6]
```diff
@@ -341,8 +361,98 @@ static int32_t PIOS_BMP280_Read(uint8_t address, uint8_t *buffer, uint8_t len)
...
+	PIOS_SPI_SetClockSpeed(bmp_dev->spi_id, PIOS_BMP_SPI_SPEED);
+
+	return 0;
+}
+
+static int32_t PIOS_BMP280_ReleaseBus(struct bmp280_dev *bmp_dev)
+{
+	PIOS_SPI_RC_PinSet(bmp_dev->spi_id, bmp_dev->spi_slave, true);
+
+	PIOS_SPI_ReleaseBus(bmp_dev->spi_id);
```
Comment (tracernz) [inline flight/PiOS/Common/pios_bmp280.c:402]: Would be nice to pass failure through even if you can't do much about it.

File: flight/PiOS/Common/pios_bmp280.c [hunk 7]
```diff
@@ -341,8 +361,98 @@ static int32_t PIOS_BMP280_Read(uint8_t address, uint8_t *buffer, uint8_t len)
...
+
+static int32_t PIOS_BMP280_SPI_Read(uint8_t address, uint8_t *buffer, uint8_t len) {
+	if (PIOS_BMP280_ClaimBus(dev) != 0)
+		return -1;
+
+	PIOS_SPI_TransferByte(dev->spi_id, 0x80 | address);
+
+	int ret = PIOS_SPI_TransferBlock(dev->spi_id, NULL, buffer, len);
+
+	PIOS_BMP280_ReleaseBus(dev);
```
Comment (tracernz) [inline flight/PiOS/Common/pios_bmp280.c:413]: As above.

File: flight/PiOS/Common/pios_bmp280.c [hunk 8]
```diff
@@ -341,8 +361,98 @@ static int32_t PIOS_BMP280_Read(uint8_t address, uint8_t *buffer, uint8_t len)
...
+	return ret;
+}
+
+static int32_t PIOS_BMP280_SPI_Write(uint8_t *buffer, uint8_t len) {
+	if (PIOS_BMP280_ClaimBus(dev) != 0)
+		return -1;
+
+	int ret = PIOS_SPI_TransferBlock(dev->spi_id, buffer, NULL, len);
+
+	PIOS_BMP280_ReleaseBus(dev);
```
Comment (tracernz) [inline flight/PiOS/Common/pios_bmp280.c:424]: As above.

**Review by tracernz — COMMENTED** [review]
Good job, looks nice. 👍

File: flight/PiOS/Common/pios_bmp280.c (hunk 5 above)
Comment (mlyle) [reply flight/PiOS/Common/pios_bmp280.c:206]: Yes, I'm not sure anything will exclude I2C.  It's just there for contingency if we're going to have a target where external BMP280 is very improbable.

**Review by tracernz — APPROVED** [review]
👍

**Pull request merged**